File: L1S/v3/code/ocaml/src/main.ml
let read_file filename =
  let ch = open_in filename in
  let len = in_channel_length ch in
  let content = really_input_string ch len in
  close_in ch;
  content

let write_file filename content =
  let ch = open_out filename in
  output_string ch content;
  close_in ch

let () =
  let args = Sys.argv in
  if Array.length args < 2 then begin
    Printf.eprintf "usage: %s <input.is> [-o <output.asm>]\n%!" args.(0);
    exit 1
  end;
  let source_file = args.(1) in
  let output_file =
    if Array.length args >= 4 && args.(2) = "-o" then args.(3)
    else Filename.remove_extension source_file ^ ".asm"
  in
  let source = read_file source_file in
  let sf = Span.make_source_file source_file source in
  (* Lex *)
  let toks, lex_diags = Lexer.tokenize source in
  if lex_diags <> [] then begin
    Diagnostic.emit sf lex_diags;
    exit 1
  end;
  (* Parse *)
  let prog = Parser.parse_program toks in
  (* Type check *)
  let tc_diags = Typeck.check_program sf prog in
  if tc_diags <> [] then begin
    Diagnostic.emit sf tc_diags;
    exit 1
  end;
  (* Build IR *)
  let ir_mod = Ir_build.build_module sf prog in
  (* Register allocate *)
  List.iter Regalloc.allocate_function ir_mod.Ir.funcs;
  (* Codegen *)
  let asm = Codegen.generate_module ir_mod in
  write_file output_file asm;
  Printf.printf "wrote %s\n%!" output_file

File: L1S/v3/code/rust/src/main.rs
mod span;
mod token;
mod lexer;
mod diag;
mod ast;
mod parser;
mod ir;
mod symbols;
mod typeck;
mod ir_build;
mod regalloc;
mod codegen;

use std::env;
use std::fs;
use std::process;

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: l1s <source.l1s> [-o output.asm]");
        process::exit(1);
    }

    let source = fs::read_to_string(&args[1]).unwrap_or_else(|e| {
        eprintln!("Error reading {}: {}", args[1], e);
        process::exit(1);
    });

    let (tokens, mut diag) = lexer::Lexer::new(&source).tokenize();

    let program = {
        let mut parser = parser::Parser::new(tokens);
        let prog = parser.parse_program();
        diag.merge(parser.diag);
        prog
    };

    if diag.has_errors() {
        diag.emit(&source);
        process::exit(1);
    }

    let mut typeck = typeck::TypeChecker::new();
    typeck.check_program(&program);
    diag.merge(typeck.diag);

    if diag.has_errors() {
        diag.emit(&source);
        process::exit(1);
    }

    let mut ir_builder = ir_build::IrBuilder::new();
    ir_builder.build(&program);

    let mut alloc = regalloc::StackAllocator::new();
    for func in &ir_builder.module.funcs {
        alloc.allocate(func.num_vregs);
    }

    let mut codegen = codegen::Codegen::new(alloc);
    let asm = codegen.generate(&ir_builder.module);

    let output_path = if args.len() > 2 && args[2] == "-o" {
        args[3].clone()
    } else {
        let input = &args[1];
        if input.ends_with(".l1s") {
            format!("{}.asm", &input[..input.len() - 4])
        } else {
            format!("{}.asm", input)
        }
    };

    fs::write(&output_path, &asm).unwrap_or_else(|e| {
        eprintln!("Error writing {}: {}", output_path, e);
        process::exit(1);
    });

    eprintln!("Wrote {}", output_path);
}
